server/etcdserver/api/v3discovery/discovery_test.go

```diff
@@ -0,0 +1,651 @@
+
+		d := &discovery{
+			lg: lg,
+			c: &clientv3.Client{
+				KV: fkv,
+			},
+			cfg:          &DiscoveryConfig{},
+			clusterToken: clusterFakeToken,
+			memberId:     tc.memberId,
+			clock:        clockwork.NewRealClock(),
+		}
+
+		clsInfo, _, _, err := d.checkCluster()
+		if err != tc.expectedError {
+			t.Errorf("case %d failed, expected error: %v, got: %v", i, tc.expectedError, err)
+		}
+
+		if err == nil {
+			if fkv.getSizeRetries != 0 || fkv.getMembersRetries != 0 {
+				t.Errorf("case %d failed, discovery client did not retry checking cluster on error, remaining etries: (%d, %d)", i, fkv.getSizeRetries, fkv.getMembersRetries)
+			}
+
+			if clsInfo.Len() != len(expectedMemberInfo) {
+				t.Errorf("case %d failed, unexpected member count, expected: %d, got: %d", i, len(expectedMemberInfo), clsInfo.Len())
+			}
+
+			for mIdx, m := range clsInfo.members {
+				if m != expectedMemberInfo[mIdx] {
+					t.Errorf("case %d failed, unexpected member[%d], expected: %v, got: %v", i, mIdx, expectedMemberInfo[mIdx], m)
+				}
+			}
+		}
+	}
+}
+
+// fakeKVForRegisterSelf is used to test registerSelf.
+type fakeKVForRegisterSelf struct {
+	*fakeBaseKV
+	t                *testing.T
+	expectedRegKey   string
+	expectedRegValue string
+	retries          int
+}
+
+// We only need to overwrite method `Put`.
+func (fkv *fakeKVForRegisterSelf) Put(ctx context.Context, key string, val string, opts ...clientv3.OpOption) (*clientv3.PutResponse, error) {
+	if key != fkv.expectedRegKey {
+		fkv.t.Errorf("unexpected register key, expected: %s, got: %s", fkv.expectedRegKey, key)
+	}
+
+	if val != fkv.expectedRegValue {
+		fkv.t.Errorf("unexpected register value, expected: %s, got: %s", fkv.expectedRegValue, val)
+	}
+
+	if fkv.retries > 0 {
+		fkv.retries--
+		// discovery client should retry on error.
+		return nil, errors.New("register self failed")
+	}
+
+	return nil, nil
+}
+
+func TestRegisterSelf(t *testing.T) {
+	cases := []struct {
+		token            string
+		memberId         types.ID
+		expectedRegKey   string
+		expectedRegValue string
+		retries          int // when retries > 0, then return an error on Put request.
+	}{
+		{
+			token:            "token1",
+			memberId:         101,
+			expectedRegKey:   "/_etcd/registry/token1/members/" + types.ID(101).String(),
+			expectedRegValue: "infra=http://127.0.0.1:2380",
+			retries:          0,
+		},
+		{
+			token:            "token2",
+			memberId:         102,
+			expectedRegKey:   "/_etcd/registry/token2/members/" + types.ID(102).String(),
+			expectedRegValue: "infra=http://127.0.0.1:2380",
+			retries:          0,
+		},
+		{
+			token:            "token3",
+			memberId:         103,
+			expectedRegKey:   "/_etcd/registry/token3/members/" + types.ID(103).String(),
+			expectedRegValue: "infra=http://127.0.0.1:2380",
+			retries:          2,
+		},
+	}
+
+	lg, err := zap.NewProduction()
+	if err != nil {
+		t.Errorf("Failed to create a logger, error: %v", err)
+	}
+
+	for i, tc := range cases {
```

Comment:
Please use subtests https://ieftimov.com/post/testing-in-go-subtests/

Reply:
Resolved & thanks.